### PR TITLE
feat: show warning about unreachable users [FS-1573]

### PR DESCRIPTION
**Pull request by V-Gira**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/FS-1573" title="FS-1573" target="_blank">Sub-taskFS-1573</a>  [web] Show system message indicating who will never receive the message
  </td></table>
  <br />
 

### Feature

Implement a warning when users from an unreachable back-end will never receive a message

Screenshot from 2023-03-17 13-54-12

#### Note

- Logic has to take into account languages with different conjugation in plural with additional internationalization strings

### CODE REVIEW

File: src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.test.tsx [hunk 1]
```diff
@@ -67,20 +75,64 @@ describe('PartialFailureToSendWarning', () => {
...
     const queued = generateUserClients(users);
     const {queryByText, container} = render(
       withTheme(<PartialFailureToSendWarning knownUsers={users} failedToSend={{queued}} />),
     );
 
     expect(queryByText('Show details')).toBeNull();
-    expect(container.textContent).toContain(`${users[0].username()} will receive your message later`);
+    expect(container.textContent).toContain(`${users[0].username()} will get your message later`);
+  });
+  it('does not show the extra info toggle if there is only a single unreachable user', () => {
```
Comment (atomrc) [inline src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.test.tsx]:
```suggestion
  });
  
  it('does not show the extra info toggle if there is only a single unreachable user', () => {
```

File: src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.test.tsx [hunk 2]
```diff
@@ -67,20 +75,64 @@ describe('PartialFailureToSendWarning', () => {
...
+    const nbUsersDomain1 = Math.floor(Math.random() * 100);
+    const nbUsersDomain2 = Math.floor(Math.random() * 100);
+    const users1 = generateQualifiedIds(nbUsersDomain1, 'domain1');
+    const users2 = generateQualifiedIds(nbUsersDomain2, 'domain2');
+
+    const failed = [...users1, ...users2];
+    const {container} = render(withTheme(<PartialFailureToSendWarning knownUsers={[]} failedToSend={{failed}} />));
+    expect(container.textContent).toContain(`${nbUsersDomain1 + nbUsersDomain2} participants didn't get your message`);
+  });
+  it('displays the number of users, named or unreachable that did not get the message across multiple domains', () => {
```
Comment (atomrc) [inline src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.test.tsx]:
```suggestion
  });
  
  it('displays the number of users, named or unreachable that did not get the message across multiple domains', () => {
```

File: src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.test.tsx [hunk 3]
```diff
@@ -67,20 +75,64 @@ describe('PartialFailureToSendWarning', () => {
       ...generateUserClients(users2),
     };
     const {container} = render(withTheme(<PartialFailureToSendWarning knownUsers={[]} failedToSend={{queued}} />));
+    expect(container.textContent).toContain(`${nbUsersDomain1 + nbUsersDomain2} participants didn't get your message`);
+  });
+  it('displays the number of unreachable users that did not get the message across multiple domains', () => {
```
Comment (atomrc) [inline src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.test.tsx]:
```suggestion
  });
  
  it('displays the number of unreachable users that did not get the message across multiple domains', () => {
```

File: src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx [hunk 4]
```diff
@@ -60,24 +61,42 @@ function generateNamedUsers(users: User[], userClients: QualifiedUserClients): P
   );
 }
 
+function generateUnreachableUsers(users: QualifiedId[]) {
+  const userCountByDomain = countBy(users, 'domain');
+  const unreachableUsers = map(userCountByDomain, (count, domain) => ({count, domain}));
+  return unreachableUsers;
```
Comment (przemvs) [inline src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx]:
```suggestion
  return map(userCountByDomain, (count, domain) => ({count, domain}));
```

File: src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx [hunk 5]
```diff
@@ -60,24 +61,42 @@ function generateNamedUsers(users: User[], userClients: QualifiedUserClients): P
...
+  const FailedToSendToOne =
     namedUsers.length === 1
-      ? {head: namedUsers[0].username(), rest: t('messageFailedToSendToOne')}
+      ? {head: namedUsers[0].username(), rest: t('messageFailedToSendWillReceiveSingular')}
       : {
-          head: t('messageFailedToSendParticipants', {count: userCount.toString()}),
-          rest: t('messageFailedToSendToSome'),
+          head: t('messageFailedToSendParticipantsFromDomainSingular', {
+            domain: unreachableUsers[0]?.domain,
+          }),
```
Comment (przemvs) [inline src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx]: Hmm.. If there we don't have domain translation will looks like that:  **1 participant from undefined**. Or I'm wrong ? 🤔

File: src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx [hunk 6]
```diff
@@ -86,18 +105,55 @@ export const PartialFailureToSendWarning = ({failedToSend, knownUsers}: Props) =
...
+            <>
+              {/* maps through the known users that will receive the message later:
+              "Alice, Bob will get your message later" */}
+              {namedUsers.length !== 0 && (
+                <p css={warning}>
+                  {namedUsers
+                    .map(user => (
+                      <Bold
+                        css={warning}
+                        data-uie-name="recipient"
```
Comment (przemvs) [inline src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx]: This name for is not consistent.

File: src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx [hunk 7]
```diff
@@ -86,18 +105,55 @@ export const PartialFailureToSendWarning = ({failedToSend, knownUsers}: Props) =
...
+                      return prev.length === 0 ? [element] : [...prev, ', ', element];
+                    }, [])}
+                  {` ${t('messageFailedToSendWillReceive')}`}
+                </p>
+              )}
+
+              {/* maps through the unreachable users that will never receive the message:
+              "3 participants from alpha.domain, 1 participant from beta.domain won't get your message" */}
+              {failed && (
+                <p css={warning} data-uie-name="failed">
```
Comment (przemvs) [inline src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx]: This name for is not consistent.

File: src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx [hunk 8]
```diff
@@ -86,18 +105,55 @@ export const PartialFailureToSendWarning = ({failedToSend, knownUsers}: Props) =
...
+                              domain: user.domain,
+                            })
+                          : t('messageFailedToSendParticipantsFromDomainSingular', {
+                              domain: user.domain,
+                            })}
+                      </Bold>
+                    ))
+                    .reduce<React.ReactNode[]>((prev, element) => {
+                      return prev.length === 0 ? [element] : [...prev, ', ', element];
+                    }, [])}
```
Comment (przemvs) [inline src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx]: This code looks similar to code on lines 113-126

File: src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx (hunk 5 above)
Comment (V-Gira) [reply src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx]: missing a conditional

File: src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx [hunk 9]
```diff
@@ -60,23 +61,41 @@ function generateNamedUsers(users: User[], userClients: QualifiedUserClients): P
...
+  if (showToggle) {
+    message.head = t('messageFailedToSendParticipants', {count: userCount.toString()});
+    message.rest = t('messageFailedToSendToSome');
+  } else if (namedUsers.length === 1) {
+    message.head = namedUsers[0].username();
+    message.rest = t('messageFailedToSendWillReceiveSingular');
+  } else if (unreachableUsers.length === 1) {
+    message.head = t('messageFailedToSendParticipantsFromDomainSingular', {domain: unreachableUsers[0].domain});
+    message.rest = t('messageFailedToSendWillNotReceiveSingular');
+  }
```
Comment (przemvs) [inline src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx:98]: Use useMemo for this :) No change obj values

Comment (atomrc) [reply src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx:98]: Oh no please, memo is really not necessary here and introduces non-needed complexity in this case

File: src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx [hunk 10]
```diff
@@ -60,23 +61,41 @@ function generateNamedUsers(users: User[], userClients: QualifiedUserClients): P
   );
 }
 
+function generateUnreachableUsers(users: QualifiedId[]) {
+  const userCountByDomain = countBy(users, 'domain');
+  return map(userCountByDomain, (count, domain) => ({count, domain}));
+}
+
+function reduceWithCommas(elements: React.ReactNode[]) {
```
Comment (atomrc) [inline src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx]: The naming is a little `off` reducing is not what the consumer wants to do. 
What they want to do is to `joinWith` and give it a custom separator :) 

```suggestion
function joinWith(elements: React.ReactNode[], separator: string) {
```

WDYT?

Comment (V-Gira) [reply src/script/components/MessagesList/Message/ContentMessage/Warnings/PartialFailureToSend/PartialFailureToSend.tsx]: wasn't happy with the naming and the custom separator sounds good :+1:

**Review by atomrc — APPROVED** [review]
Solid PR 💪 😎

**Pull request merged**